docs/resources/server/damagetypes.md

```diff
@@ -0,0 +1,127 @@
+
+```json5
+{
+  // The death message id of the damage type. The full death message translation key will be
+  // "death.examplemod.example" (with swapped-out mod ids and names).
+  "message_id": "example",
+  // Whether this damage type's damage amount scales with difficulty or not. Valid values are:
+  // - "never": The damage value remains the same on any difficulty. Common for player-caused damage types.
+  // - "when_caused_by_living_non_player": The damage value is scaled if the entity is caused by a
+  //   living entity of some sort, including indirectly (e.g. an arrow shot by a skeleton), that is not a player.
+  // - "always": The damage value is always scaled. Commonly used by explosion-like damage.
+  "scaling": "when_caused_by_living_non_player",
+  // The amount of exhaustion caused by receiving this kind of damage.
+  "exhaustion": 0.1,
+  // The damage effects (currently only sound effects) that are applied when receiving this kind of damage. Optional.
+  // Valid values are "hurt" (default), "thorns", "drowning", "burning", "poking", and "freezing".
+  "effects": "hurt",
+  // The death message type. Determines how the death message is built. Optional.
+  // Valid values are "default" (default), "fall_variants", and "intentional_game_design".
+  "death_message_type": "default"
+}
+```
+
+:::tip
+The `scaling`, `effects` and `death_message_type` fields are internally controlled by the enums `DamageScaling`, `DamageEffects` and `DeathMessageType`, respectively. These enums can be [extended][extenum] to add custom values if needed.
+:::
+
+The same format is also used for vanilla's damage types, and pack developers can change these values if needed.
+ 
+## Creating and Using Damage Sources
+
+`DamageSource`s are usually created on the fly when `Entity#hurt` is called. Be aware that since damage types are a [datapack registry][dr], you will need a `RegistryAccess` to query them, which can be obtained via `Level#registryAccess`. To create a `DamageSource`, call the `DamageSource` constructor with up to four parameters:
+
+```java
+DamageSource damageSource = new DamageSource(
+        // The damage type holder to use. Query from the registry. This is the only required parameter.
+        registryAccess.registryOrThrow(Registries.DAMAGE_TYPE).getHolderOrThrow(EXAMPLE_DAMAGE),
+        // The entity causing the damage. This isn't always applicable (e.g. when falling out of the world)
+        // and may therefore be null. Optional, defaults to null.
+        null,
+        // The direct entity. For example, if a skeleton shot you, the skeleton would be the causing entity
+        // (= the parameter above), and the arrow would be the direct entity (= this parameter). Similar to
+        // the causing entity, this isn't always applicable and therefore nullable. Optional, defaults to null.
+        null,
```

Comment:
The descriptions for these two are flipped.

Reply:
No, that's correct. It's causing entity (e.g. skeleton), and then direct entity (e.g. arrow).

Reply:
This is the constructor you are using. You can check yourself.

```java
public DamageSource(Holder<DamageType> type, @Nullable Entity directEntity, @Nullable Entity causingEntity, @Nullable Vec3 damageSourcePosition) {
        this.type = type;
        this.causingEntity = causingEntity;
        this.directEntity = directEntity;
        this.damageSourcePosition = damageSourcePosition;
    }
```